enhancements/authentication/service-ca-cert-generation-for-statefulset-pods.md

```diff
@@ -0,0 +1,272 @@
+Second best is to rely on known properties of the environment,
+i.e. to assume that 127.0.0.1 is localhost-only
+and to use a non-TLS connection on that address.
+
+If a pod _had_ to use TLS for localhost for some reason, the closest to a right way to do that
+(relying on TLS to verify that only loopback connections happen) is to:
+- Create a temporary CA (specific to the individual pod)
+- Generate, and have that CA sign, a localhost certificate (specific to the individual pod)
+- Set up the CA certificate as a trusted root
+- (Irreversibly erase the CA private key to minimize risk)
+- Set up the pod with the generated localhost certificate+private key.
+- (This doesn’t benefit from service-ca involvement at all.)
+
+Similarly, a single-purpose CA could be set up for NodeIP services
+(if it makes sense to connect to node IP addresses directly at all)
+so that the client connects to the desired NodeIP service
+instead of to any NodeIP service.
+
+### Risks and Mitigations
+
+The only mitigations for the risks below are not to use this feature,
+(and prevent untrusted users from using it — using an admission controller?)
+so that the wildcard certificate is not created at all.
+
+#### Unexpected DNS Collisions in the Future
+
+https://github.com/kubernetes/dns/blob/master/docs/specification.md
+only specifies what DNS records must exist, not what must not exist,
+so the generated wildcard certificate could match unwanted DNS entries.
+
+The DNS specification does not even guarantee that different namespaces
+will use different DNS subtrees.
+
+Given the existing naming conventions of in-cluster DNS,
+such unexpected collisions seem somewhat unlikely,
+but can’t be ruled out in advance.
+
+#### No Individual Pod Identity
+
+The wildcard certificate does not protect the connections to the StatefulSet
+against re-routing to unintended members of that StatefulSet
+(or to a different StatefulSet attached to the same headless service).
+
+In addition, any application that uses a different set of certificates to achieve that goal
+**must not accept the Service CA as a valid root of trust for connections intended
+for individual StatefulSet pods**.
+
+## Design Details
+
+### Open Questions
+
+N/A
+
+### Test Plan
+
+Unit and e2e tests similar to the existing service certificate tests, both in structure and scope.
+
+Ensure that upgrades (or alternatively any controller re-syncs)
+cause re-generation of certificates for headless services
+to include the wildcard subjects, if they are missing.
+
+### Graduation Criteria
+
+N/A, just a code change to the Service CA.
+
+TBD: End-user documentation.
+
+### Upgrade / Downgrade Strategy
+
+Upgrading to a version that adds the feature will trigger a regeneration of secrets
+requested for headless services.
+The previously-generated certificates will continue to be valid for their original lifetime,
+and previously-deployed components can continue to use them as before.
```

Comment:
is this supposed to be tested manually by QE?

Reply:
The upgrade case is unit-tested ( https://github.com/openshift/service-ca-operator/pull/149 `"secret already contains pre-headless cert data, update to headless"` ).

That certificates continue to work until regenerated feels trivial to me (is that mistaken?), and as long as no OpenShift components start relying on the wildcard certificates in the 4.8 cycle, not likely to break anything in practice anyway AFAICS.

Reply:
> is this supposed to be tested manually by QE?

@barleyer fyi ^